### PR TITLE
Generate some ml type definitions

### PR DESCRIPTION
**Pull request by Nadrieril**
Progress on #215. This auto-generates a number of ml definitions from the rust ones. I've left the visitor definitions as-is; that feels incomplete, I hope to find a cleaner solution when more of the types are auto-generated.

Nice bonus: this copies the doc-comments present on the rust side.

### CODE REVIEW

File: charon-ml/src/Expressions.ml [hunk 1]
```diff
@@ -1,27 +1,75 @@
+(** WARNING: this file is partially auto-generated. Do not edit `src/Expressions.ml`
+    by hand. Edit `templaces/Expressions.ml` instead, or improve the code
```
Comment (sonmarcho) [inline charon-ml/src/Expressions.ml:2]: "templaces"

File: charon-ml/src/Expressions.ml [hunk 2]
```diff
@@ -1,27 +1,75 @@
...
 type var_id = VarId.id [@@deriving show, ord]
 
+(** An assumed function identifier, identifying a function coming from a
+ standard library. *)
 type assumed_fun_id =
-  | BoxNew
+  | BoxNew  (** `alloc::boxed::Box::new` *)
   | BoxFree
+      (** `alloc::alloc::box_free`
+ This is actually an unsafe function, but the rust compiler sometimes
```
Comment (sonmarcho) [inline charon-ml/src/Expressions.ml:29]: We should add indentation in multi-line comments

**Comment by sonmarcho** [comment]
I hadn't thought of the fact that doing this would allow to port the documentation from the Rust side to the OCaml side: this is extremely useful. The only minor change I would like to do is a better indentation of the multi-line comments (as you can see, having no indentation is a bit ugly).

File: charon-ml/src/Expressions.ml (hunk 2 above)
Comment (Nadrieril) [reply charon-ml/src/Expressions.ml:29]: The `--wrap-comments` ocamlformat option doesn't work on doc-comments and destroys the formatting of e.g. code blocks

Comment (Nadrieril) [reply charon-ml/src/Expressions.ml:29]: I can still guess the indent level, that'd be better than the current state.

Comment (Nadrieril) [reply charon-ml/src/Expressions.ml:29]: This worked great (https://github.com/AeneasVerif/charon/pull/307)

**Pull request merged**